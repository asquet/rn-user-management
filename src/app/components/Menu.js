import React from 'react';
import { View, Text, TouchableOpacity } from 'react-native';
import Drawer from 'react-native-drawer';
import { menuStyles } from '../stylesheet';

export default class Menu extends React.Component {

  static get childContextTypes() {
    return {
      menuOpen: React.PropTypes.func,
      menuClose: React.PropTypes.func,
    };
  }

  constructor(props) {
    super(props);

    this.drawerRef = null;
  }

  getChildContext() {
    return {
      menuOpen: this.openMenu.bind(this),
      menuClose: this.closeMenu.bind(this),
    };
  }

  closeMenu() {
    this.drawerRef.close();
  }

  openMenu() {
    this.drawerRef.open();
  }

  render() {
    const {
      menuItems,
      children,
    } = this.props;
    return (
      <Drawer
        captureGestures
        ref={(ref) => {
          this.drawerRef = ref;
        }}
        tapToClose
        openDrawerOffset={0.2} // 20% gap on the right side of drawer
        panCloseMask={0.2}
        closedDrawerOffset={-3}
        tweenHandler={ratio => ({
          main: { opacity: (2 - ratio) / 2 },
        })}
        styles={{
          main: menuStyles.drawerMain,
          drawer: menuStyles.drawerSelf,
          drawerOverlay: menuStyles.drawerSelfOverlay,
          mainOverlay: menuStyles.drawerMainOverlay,
        }}
        content={
          <View style={menuStyles.container}>
            <Text style={menuStyles.header}>Menu</Text>
            <View style={menuStyles.listContainer}>
              {
                menuItems.map(item => (
                  <TouchableOpacity
                    onPress={item.onPress}
                    key={item.text}
                    style={menuStyles.menuItemButton}
                  >
                    <Text style={menuStyles.menuItemText}>{item.text}</Text>
                  </TouchableOpacity>
                ))
              }
            </View>
          </View>
        }
      >
        {children}
      </Drawer>
    );
  }

}

Menu.propTypes = {
  menuItems: React.PropTypes.arrayOf(React.PropTypes.shape({
    text: React.PropTypes.string,
    onPress: React.PropTypes.func,
  })).isRequired,
  children: React.PropTypes.node.isRequired,
};

Menu.withMenuAwareness = function genWithMenuAwareness(Cmp) {
  class withMenuAwareness extends React.Component {
    render() {
      return (
        <Cmp {...this.props} menuOpen={this.context.menuOpen} menuClose={this.context.menuClose} />
      );
    }
  }

  withMenuAwareness.contextTypes = {
    menuOpen: React.PropTypes.func.isRequired,
    menuClose: React.PropTypes.func.isRequired,
  };

  return withMenuAwareness;
};
