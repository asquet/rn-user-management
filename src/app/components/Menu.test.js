import { Text } from 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import Menu from './Menu';

// Note: test renderer must be required after react-native.

describe('Menu', () => {
  it('renders correctly', () => {
    const menuItems = [
      {
        text: 'p1',
        onPress: () => {
        },
      },
    ];
    const tree = renderer.create(
      <Menu menuItems={menuItems}><Text>OK</Text></Menu>,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
