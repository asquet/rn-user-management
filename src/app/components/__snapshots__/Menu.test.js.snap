// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    elevation={0}
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      Array [
        Object {
          "borderWidth": 0,
          "left": -603,
          "position": "absolute",
          "shadowColor": "#000",
          "shadowOpacity": 0.8,
          "shadowRadius": 3,
          "top": 0,
        },
        Object {
          "height": 1334,
          "width": 600,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "white",
          "flex": 1,
          "padding": 10,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 25,
            "fontWeight": "bold",
            "marginBottom": 20,
          }
        }
      >
        Menu
      </Text>
      <View
        style={
          Object {
            "flex": 0,
            "flexDirection": "column",
            "justifyContent": "flex-start",
          }
        }
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginBottom": 5,
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "fontSize": 20,
              }
            }
          >
            p1
          </Text>
        </View>
      </View>
    </View>
    <View
      pointerEvents="auto"
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {},
        ]
      }
    />
  </View>
  <View
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      Array [
        Object {
          "borderWidth": 0,
          "left": -3,
          "paddingLeft": 3,
          "position": "absolute",
          "top": 0,
        },
        Object {
          "height": 1334,
          "width": 753,
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      OK
    </Text>
    <View
      pointerEvents="none"
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {},
        ]
      }
    />
  </View>
</View>
`;
